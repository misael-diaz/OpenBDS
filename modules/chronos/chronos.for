!
!   source: chronos.for
!   author: misael-diaz
!   date:   2021-05-31
!
!
!   Synopsis:
!   Defines timer classes.
!
!
!   Copyright (C) 2021 Misael Diaz-Maldonado
!
!   This program is free software: you can redistribute it and/or modify
!   it under the terms of the GNU General Public License as published by
!   the Free Software Foundation, either version 3 of the License, or
!   (at your option) any later version.
!
!   This program is distributed in the hope that it will be useful,
!   but WITHOUT ANY WARRANTY; without even the implied warranty of
!   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!   GNU General Public License for more details.
!
!   You should have received a copy of the GNU General Public License
!   along with this program.  If not, see <http://www.gnu.org/licenses/>.

module chronos
    use, intrinsic :: iso_fortran_env, only: int32, int64, real64
    implicit none

    type, public :: chronom
        private
        integer(kind = int64):: b               ! begin time
        integer(kind = int64):: e               ! end time
        real(kind = real64):: lapse             ! elapsed-time
        real(kind = real64):: period            ! clock period
        contains
            procedure, public :: tic => start_method
            procedure, public :: toc => stop_method
            procedure, public :: etime => elapsed_time_method
    end type

    interface chronom
        module procedure constructor
    end interface

    private
    contains
        function constructor() result(chron)
            type(chronom):: chron
            
            call clockspecs(chron)
            
            return
        end function

        subroutine clockspecs(chronometer)
            ! Synopsis:
            ! Queries the system clock to set the period of the chronometer
            ! and assigns default-values to its data members (or fields).

            class(chronom):: chronometer
            real(kind = real64):: clock_period
            integer(kind = int64):: clock_rate


            call system_clock(count_rate = clock_rate)
            clock_period = 1.0_real64 / real(clock_rate, kind = real64)


            chronometer % b      = 0_int64
            chronometer % e      = 0_int64
            chronometer % lapse  = 0.0_real64
            chronometer % period = clock_period

            return
        end subroutine clockspecs

        subroutine start_method(this)
            ! Synopsis:
            ! Starts the chronometer.

            class(chronom), target :: this
            integer(kind = int64), pointer :: p_beg

            p_beg  => this % b
            call system_clock(p_beg)

            return
        end subroutine

        subroutine stop_method(this)
            ! Synopsis:
            ! Stops the chronometer.

            class(chronom), target :: this
            integer(kind = int64), pointer :: p_end

            p_end => this % e
            call system_clock(p_end)

            return
        end subroutine

        function elapsed_time_method(this) result(etime)
            ! Synopsis:
            ! Returns the elapsed-time in milliseconds.

            class(chronom), target :: this
            
            integer(kind = int64), pointer :: p_beg
            integer(kind = int64), pointer :: p_end

            real(kind = real64):: etime
            real(kind = real64), pointer :: p_etime
            real(kind = real64), pointer :: p_period


            p_beg    => this % b
            p_end    => this % e
            p_etime  => this % lapse
            p_period => this % period


            p_etime = real( (p_end - p_beg), kind = real64 ) * p_period
            p_etime = 1.0e+3_real64 * p_etime
            etime   = p_etime

            return
        end function
end module
