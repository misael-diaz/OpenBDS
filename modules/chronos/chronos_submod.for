!
!   source: chronos.for
!   author: misael-diaz
!   date:   2021-05-31
!
!
!   Synopsis:
!   Implements the procedures of the timer classes.
!
!
!   Copyright (C) 2021 Misael Diaz-Maldonado
!
!   This program is free software: you can redistribute it and/or modify
!   it under the terms of the GNU General Public License as published by
!   the Free Software Foundation, either version 3 of the License, or
!   (at your option) any later version.
!
!   This program is distributed in the hope that it will be useful,
!   but WITHOUT ANY WARRANTY; without even the implied warranty of
!   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!   GNU General Public License for more details.
!
!   You should have received a copy of the GNU General Public License
!   along with this program.  If not, see <http://www.gnu.org/licenses/>.

submodule (chronos) chronos_implementations
    use, intrinsic :: iso_fortran_env, only: int64, real64
    implicit none
    contains


        module function constructor () result(chron)
            type(chronom):: chron
            
            call clockspecs(chron)
            
            return
        end function


        module subroutine clockspecs (chronometer)
            ! Synopsis:
            ! Queries the system clock to set the period of the chronometer
            ! and assigns default-values to its data members (or fields).

            type(chronom), intent(inout) :: chronometer
            real(kind = real64):: clock_period
            integer(kind = int64):: clock_rate


            call system_clock(count_rate = clock_rate)
            clock_period = 1.0_real64 / real(clock_rate, kind = real64)


            chronometer % b      = 0_int64
            chronometer % e      = 0_int64
            chronometer % lapse  = 0.0_real64
            chronometer % period = clock_period

            return
        end subroutine clockspecs


        module subroutine start_method (self)
            ! Synopsis:
            ! Starts the chronometer.

            class(chronom), intent(inout), target :: self
            integer(kind = int64), pointer :: p_beg

            p_beg  => self % b
            call system_clock(p_beg)

            return
        end subroutine


        module subroutine stop_method (self)
            ! Synopsis:
            ! Stops the chronometer.

            class(chronom), intent(inout), target :: self
            integer(kind = int64), pointer :: p_end

            p_end => self % e
            call system_clock(p_end)

            return
        end subroutine


        module function elapsed_time_method (self) result(etime)
            ! Synopsis:
            ! Returns the elapsed-time in milliseconds.

            class(chronom), intent(inout) :: self
            real(kind = real64):: etime

            associate (begin => self % b,        end => self % e, &
                     & lapse => self % lapse, period => self % period)
                lapse = real( (end - begin), kind = real64 ) * period
                lapse = 1.0e+3_real64 * lapse
                etime = lapse
            end associate

            return
        end function
end submodule
