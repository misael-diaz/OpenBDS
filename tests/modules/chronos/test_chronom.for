!
!   source: test_chronom.for
!   author: misael-diaz
!   date:   2021-05-31
!
!
!   Synopsis:
!   Tests the chronometer class by delaying the code with expensive 64-bit
!   integer arithmetic operations. Its output is comparable to that produced
!   by the Linux time utility.
!
!
!   Copyright (C) 2021 Misael Diaz-Maldonado
!
!   This program is free software: you can redistribute it and/or modify
!   it under the terms of the GNU General Public License as published by
!   the Free Software Foundation, either version 3 of the License, or
!   (at your option) any later version.
!
!   This program is distributed in the hope that it will be useful,
!   but WITHOUT ANY WARRANTY; without even the implied warranty of
!   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!   GNU General Public License for more details.
!
!   You should have received a copy of the GNU General Public License
!   along with this program.  If not, see <http://www.gnu.org/licenses/>.

module test_chronom_module
    use, intrinsic :: iso_fortran_env, only: int64
    implicit none
    private
    public :: test_outcome
    contains
        subroutine test_outcome(i, j)
            ! Synopsis:
            ! Performs a simple equality test.

            integer(kind = int64) :: i
            integer(kind = int64) :: j

            if (i == j) then
                print *, 'equality test: passed'
            else
                print *, 'equality test: failed'
            end if

            return
        end subroutine
end module

program test_chronom
    use, intrinsic :: iso_fortran_env, only: int32, int64
    use chronos, only: chronom
    use test_chronom_module, only: test_outcome
    implicit none

    type(chronom) :: stopwatch

    integer(kind = int32):: i 
    integer(kind = int32):: j
    integer(kind = int64):: total
    integer(kind = int64):: counter = 0_int64
    integer(kind = int32), parameter :: n = 65536


    stopwatch = chronom()               ! instantiates chronometer object
    call stopwatch % tic()


    i = 0
    do while(i /= n)
        j = 0
        do while(j /= n)
            counter = counter + 1_int64         ! delays code
            j = j + 1
        end do
        i = i + 1
    end do
    
    
    call stopwatch % toc()

    ! displays test outcome for the user just to be thorough
    total = int(n, kind = int64) * int(n, kind = int64)
    call test_outcome(total, counter)
    print *, 'total: ', total, counter
    print *, 'elapsed-time (millis): ', stopwatch % etime()
    
end program


! Comments:
! The main objective is to perform a runtime test. Since the real work is
! delegated to the instrinsic procedure system_clock(), we do not bother
! to check the implementation (which can be boiled down to taking a time
! difference and multiplying a conversion factor). Nevetheless, the
! elapsed-time shown by the program is consistent with the real time.
!
! We compare the elapsed-time generated by the chronometer with that
! produced by the Linux time utility. The output of this test can be
! found in: test_chronom.out. (Note that these differ due to the
! overhead of writing to standard output not accounted for by the
! choronometer.)
!
! I could not help myself but to test for the correctness of the integer
! arithmetic computations even though I know it's not strictly necessary.
