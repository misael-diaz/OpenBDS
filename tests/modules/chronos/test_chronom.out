

Output shows that the real-time reported by time is consistent
with that computed by the chronometer.

$ time ./test-chronometer-class 
 total:            4294967296           4294967296
 test: passed
 elapsed-time (millis):    6689.0567830000009     

 real	0m6.691s
 user	0m6.637s
 sys	0m0.044s


